dwave/preprocessing/presolve/cypresolve.pyx

```diff
@@ -137,7 +137,8 @@ cdef class cyPresolver:
         cdef bint changes = False
 
         try:
-            changes = self.cpppresolver.normalize()
+            with nogil:
+                changes = self.cpppresolver.normalize()
```

Comment:
Is this safe?

Unless I'm missing something, `.normalize()` and `.presolve()` are not thread-safe because they are potentially modifying the (shared) model inplace, and I don't see any synchronization mechanism implemented.

Reply:
Sure, I can add a lock or at least a note saying that they are not threadsafe. The intent here is not to multithread `normalize()` and `presolve()` but to allow other threads to proceed while presolve is inside the C++ block.

Reply:
But the problem is if other threads are accessing the same CQM model (that's being modified here).

Reply:
Each presolver owns its own model. So
```python
cqm = dimod.CQM()

presolvers = [Presolver(cqm) for _ in range(num_presolvers)]

for i in prange(num_presolvers):
    presolver[i].apply()
```
is safe. But
```
cqm = dimod.CQM()

presolver = Presolver(cqm)

for _ in prange(num_threads):
    presolver.apply()
```
is not.

Reply:
Ok, if they make a copy of the cqm, then cqm is safe. But there's still the model in presolver that's unprotected from modifications in parallel, so adding at least a note about that would be good. A lock would be nice, but at this point, as you note, would probably be an overkill.

Reply:
Good timing, I just added a lock in 7b3c4d6 via C++ mutex. I considered making the underlying C++ class threadsafe via mutex but for now I think it makes sense to lock it from the outside.
The following script
```python
import concurrent.futures
import time

import dimod
import mip

from dwave.preprocessing import Presolver

# m = mip.Model()
# m.read("neos-4413714-turia.mps")
# m.write("neos-4413714-turia.lp")

cqm = dimod.lp.load("neos-4413714-turia.lp")

presolver = Presolver(cqm)
presolver.load_default_presolvers()
presolver.apply()  # apply once so subsequent calls to presolve will all take approx the same time

# establish the time of a single call to presolve
t = time.perf_counter()
presolver.presolve()
t_single = time.perf_counter() - t

num_threads = 4

with concurrent.futures.ThreadPoolExecutor(max_workers=num_threads) as executor:
    t = time.perf_counter()
    for i in range(num_threads):
        concurrent.futures.wait([executor.submit(presolver.presolve)])
    t_serial = time.perf_counter() - t

    t = time.perf_counter()
    jobs = [executor.submit(presolver.presolve) for i in range(num_threads)]
    concurrent.futures.wait(jobs)
    t_parallel = time.perf_counter() - t

print("   num_threads:", num_threads)
print(f"  time single: {t_single:.4f}")
print(f"  time serial: {t_serial:.4f} ({t_serial / t_single:.2f}x time single)")
print(f"time parallel: {t_parallel:.4f} ({t_parallel / t_single:.2f}x time single)")
```
now returns 
```
   num_threads: 4
  time single: 1.4978
  time serial: 6.1584 (4.11x time single)
time parallel: 6.1010 (4.07x time single)
```
as expected.

Reply:
I'll add a few unittests for thread safety shortly.